doc/design/state-machine.md

```diff
@@ -0,0 +1,185 @@
+
+__e<sub>0</sub>__:
+
+Condition: all cells allocated to the AG are in Free or Reserved state (we assume the cell allocation algorithm has ensured that this AG’s priority is higher than that every Reserved cell, i.e., that of the Preempting AG).
+
+Operation: set the cells to Used; set the Preempting AG on the Reserved cells to Pending.
+
+__e<sub>1</sub>__:
+
+Condition: there is at least one cell allocated to the AG in Used or Reserving state (we assume the cell allocation algorithm has ensured that this AG’s priority is higher than that of every Used and Reserving cell).
+
+Operation: set all the Used cells to Reserving state (e<sub>2</sub> in cell state machine); overwrite the Reserved (e<sub>6</sub> in cell state machine) and Reserving cells (e<sub>3</sub> in cell state machine); set the Free cell to Reserved (e<sub>5</sub> in cell state machine).
+
+__e<sub>2</sub>__:
+
+Condition: all the cells allocated to the AG are Reserved.
+
+Operation: set all the cells to Used (e<sub>8</sub> in cell state machine).
+
+__e<sub>3</sub>__:
+
+Condition: all pods deleted.
+
+Operation: set all the cells allocated to the AG in Free state (e<sub>1</sub> in cell state machine).
+
+__e<sub>4</sub>__:
+
+Condition: all pods deleted.
+
+Operation: none.
+
+__e<sub>5</sub>__:
+
+Condition: a Reserving or Reserved cell in this AG is being overwritten (e<sub>3</sub> or e<sub>6</sub> in cell state machine).
+
+Operation: set all the other Reserving cells to Used (e<sub>4</sub> in cell state machine), Reserved cells to Free (e<sub>7</sub> in cell state machine).
+
+__e<sub>6</sub>__:
+
+Condition: a cell from Used to Reserving (e<sub>2</sub> in cell state machine)
+
+Operation: none.
+
+__e<sub>7</sub>__:
+
+Condition: AG deleted.
+
+Operation: set the Reserved cells to Free (e<sub>7</sub> in cell state machine), set the Reserving cells to Used (e<sub>4</sub> in cell state machine).
+
+## Cell State Machine
+
+Cell is the resource unit in HiveD. The figure below shows the state machine of cell. Note that here cells are _lowest-level physical cells_ (we record states only in these cells).
+
+![AG](cell-state-machine.png)
+
+### States
+
+__Free__: the cell is not allocated to any AG.
+
+__Used__: the cell is allocated to an AG, and the AG is running on this cell.
+
+__Reserved__: the cell is allocated to an AG, but there is no AG running on this cell (e.g., the AG is Preempting).
+
+__Reserving__: the cell is allocated to an AG, but another AG is still running on the cell (e.g., the first AG is Preempting the running AG).
+
+Note that the reservation of cells (Reserved and Reserving states) is not necessarily designed for preemptions (i.e., reserving resources for the Preempting AGs). In the future it is possible that we extend this mechanism to support other reservation mechanisms.
```

Comment:
to support other features that needs reservation, such as reservation during waiting to achieve strict FIFO and fairness for larger AG.